### ISSUE TITLE
can I fix the GPU usage during training？

### ISSUE THREAD

**Issue by Monlter** [ultralytics/ultralytics#21016]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, can I fix the GPU usage during training? In other words, I can prevent the GPU memory usage from fluctuating. This can prevent my program from failing to apply for GPU resources when other nodes use too much GPU resources during training.
The main problem is that when I am at a certain stage of the training process, the GPU memory occupied by the training code suddenly decreases, and other programs may take the opportunity to use more GPU memory. When the training code needs to apply for more GPU memory, the training code crashes due to lack of extra GPU memory. This problem bothers me a lot.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Monlter, thank you for reaching out and using Ultralytics 🚀! This is an automated response to help guide you while an Ultralytics engineer reviews your question.

For resource management and GPU usage customization, you may find helpful information in our [Docs](https://docs.ultralytics.com/)—especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) sections. Many common questions about training and hardware compatibility are addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

If you are looking for best practices in custom training or troubleshooting, please share as much detail as possible—such as environment specs, training logs, and configuration files. Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for support and discussion:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Make sure you are running the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to see if your issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can run in any of these verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the badge above is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. Our CI system verifies the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

An Ultralytics engineer will review your question soon. Thank you for your patience and for being part of the Ultralytics community! 🧡

**Comment by glenn-jocher** [comment]
You can stabilize GPU memory usage by pre-allocating memory at the start of training. Try setting the `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:False` environment variable before training, or use `torch.cuda.set_per_process_memory_fraction(0.8)` in your script to reserve a fixed portion of GPU memory. Additionally, you can maximize your batch size using `batch=-1` which will automatically determine the optimal batch size for your GPU and help maintain more consistent memory usage throughout training.

**Comment by Monlter** [comment]
Thank you for your reply. The above solutions do not solve the problem I raised very well. I want to use a fixed GPU memory during my training process (which can be the maximum memory required by the program and data) without any fluctuations, so that other programs will occupy my memory resources due to fluctuations in memory during my training process.

**Comment by Monlter** [comment]
For practical situations, during the training process (currently only one program is being trained), the gpu memory I use is sometimes 20000MB and sometimes 5000MB. I want to keep the gpu memory I use fixed at 20000MB throughout the training process

**Comment by glenn-jocher** [comment]
You can pre-allocate a fixed amount of GPU memory at the start of training by creating a dummy tensor to reserve the maximum memory needed. Try this approach:

```python
import torch
from ultralytics import YOLO

# Pre-allocate GPU memory (adjust size as needed)
device = torch.device('cuda:0')
dummy_tensor = torch.empty(int(20000 * 1024**2 // 4), dtype=torch.float32, device=device)  # ~20GB
dummy_tensor = dummy_tensor.view(-1)  # Keep reference to prevent garbage collection

# Now train your model
model = YOLO('yolo11n.pt')
results = model.train(data='your_data.yaml', epochs=100)

# Keep dummy_tensor in scope throughout training
```

This reserves the memory upfront and prevents other processes from claiming it when your training temporarily uses less memory. The tensor stays allocated for the entire training duration, maintaining consistent memory usage.
